opt: panic: caused by could not find overload for binary expression concat

Seems related to #34930 , but looks like this didn't get fixed by its fix. @andy-kimball for triage.

```
[15:30:1]% ./cockroach demo                                                                                           jordan@Termato:~/repo/src/github.com/cockroachdb/cockroach on master -> origin/master
#
# Welcome to the CockroachDB demo database!
#
# You are connected to a temporary, in-memory CockroachDB
# instance. Your changes will not be saved!
#
# Web UI: http://127.0.0.1:52061
#
# Server version: CockroachDB CCL v19.1.0-beta.20190225-202-g53da5aa7bd (x86_64-apple-darwin18.2.0, built 2019/02/27 20:13:32, go1.11) (same version as client)
# Cluster ID: c51fc3f6-1b09-41b1-b711-4ae776582955
#
# Enter \? for a brief introduction.
#
root@127.0.0.1:52060/defaultdb> CREATE TABLE table0 (col0 INT8 NOT NULL, col1 STRING, col2 OID[], col3 DATE NOT NULL, col4 DATE, col5 TIMESTAMP NOT NULL, FAMILY fam0 (col1, col5, col0, col4, col3, col2), PRIMARY KEY (col0, col5 ASC, col4 ASC, col3, col1 DESC), INDEX (col3 DESC, col4 DESC, col5, col0, col1));CREATE TABLE table1 (col0 TIME NULL, col1 INT8[] NOT NULL, FAMILY fam0 (col1, col0), PRIMARY KEY (col0 ASC), UNIQUE (col0 ASC), INDEX (col0), UNIQUE (col0), INDEX (col0 DESC), UNIQUE (col0 ASC));CREATE TABLE table2 (col0 INT8[] NOT NULL, col1 INT8 NULL, col2 BOOL, col3 VARBIT, col4 INTERVAL NOT NULL, col5 BYTES, col6 INET NOT NULL, col7 STRING, col8 TIMESTAMPTZ NOT NULL, col9 TIMESTAMPTZ, col10 STRING COLLATE da NULL, col11 BOOL NOT NULL, col12 TIMESTAMPTZ, col13 BOOL NULL, col14 NAME NOT NULL, col15 OID[], col16 OID[], col17 UUID, col18 INET NULL, FAMILY fam0 (col11), FAMILY fam1 (col5), FAMILY fam2 (col9, col2), FAMILY fam3 (col14), FAMILY fam4 (col16), FAMILY fam5 (col15), FAMILY fam6 (col3, col8, col0, col4), FAMILY fam7 (col13), FAMILY fam8 (col1), FAMILY fam9 (col10), FAMILY fam10 (col7, col6, col17, col12), FAMILY fam11 (col18), PRIMARY KEY (col11 DESC), UNIQUE (col1 DESC, col4 DESC), UNIQUE (col14, col6 ASC, col2, col4 DESC, col18 DESC), INDEX (col11 DESC, col12), INDEX (col11, col14 DESC), INDEX (col2, col10, col6 DESC, col8 ASC, col18, col7 ASC));CREATE TABLE table3 (col0 INET, col1 DATE NOT NULL, col2 NAME, col3 BOOL NULL, col4 DATE NOT NULL, col5 INTERVAL, col6 TIMESTAMP NOT NULL, col7 INT8[] NULL, col8 BYTES, col9 TIME NULL, col10 INT8 NULL, col11 INT8[] NOT NULL, col12 STRING COLLATE en NOT NULL, col13 TIME NOT NULL, col14 DECIMAL, FAMILY fam0 (col12), FAMILY fam1 (col14), FAMILY fam2 (col1), FAMILY fam3 (col7), FAMILY fam4 (col11), FAMILY fam5 (col3), FAMILY fam6 (col4), FAMILY fam7 (col6), FAMILY fam8 (col10), FAMILY fam9 (col2), FAMILY fam10 (col8), FAMILY fam11 (col9), FAMILY fam12 (col0), FAMILY fam13 (col5), FAMILY fam14 (col13), PRIMARY KEY (col12 DESC), UNIQUE (col5 ASC, col13 DESC, col2 ASC, col9 ASC, col4, col0, col1 ASC, col8 ASC, col12, col3 DESC, col6), INDEX (col2, col14, col4 DESC, col1 DESC, col9 DESC), UNIQUE (col0 ASC), UNIQUE (col9 ASC, col1, col10 DESC, col4 ASC, col12 ASC, col3 DESC, col6 ASC), UNIQUE (col13 DESC, col6));CREATE TABLE table4 (col0 FLOAT8, col1 STRING COLLATE en, FAMILY fam0 (col1, col0), PRIMARY KEY (col1 ASC), INDEX (col1), UNIQUE (col0 ASC, col1), UNIQUE (col1, col0), UNIQUE (col1 ASC), INDEX (col0), UNIQUE (col0, col1 ASC), INDEX (col0 DESC));CREATE TABLE table5 (col0 OID NOT NULL, FAMILY fam0 (col0), PRIMARY KEY (col0 DESC), INDEX (col0));CREATE TABLE table6 (col0 TIME NOT NULL, col1 OID NOT NULL, col2 FLOAT8 NULL, col3 UUID NOT NULL, col4 FLOAT8 NOT NULL, col5 OID NOT NULL, col6 BOOL NOT NULL, FAMILY fam0 (col6, col1, col4, col0, col5, col2, col3), PRIMARY KEY (col1 ASC, col0 ASC), INDEX (col3 ASC, col0 ASC, col5 ASC, col1), UNIQUE (col1 ASC, col6 DESC), INDEX (col4 DESC, col1 DESC, col6, col0 DESC, col5, col2, col3 ASC), INDEX (col2 DESC, col0, col6 ASC), UNIQUE (col5 DESC, col0 DESC), INDEX (col5 ASC, col2, col1), INDEX (col4 DESC, col6, col5, col3, col1), INDEX (col2 ASC), UNIQUE (col3 DESC, col4 ASC, col0, col1 ASC, col6 ASC));CREATE TABLE table7 (col0 TIMESTAMP, col1 VARBIT NOT NULL, col2 STRING COLLATE da NOT NULL, col3 INT8[] NULL, col4 INT8 NOT NULL, col5 INT8[] NULL, col6 VARBIT NOT NULL, col7 OID, col8 INTERVAL NULL, col9 TIME NOT NULL, col10 BYTES, col11 BYTES, col12 DECIMAL NOT NULL, FAMILY fam0 (col2), FAMILY fam1 (col5, col10), FAMILY fam2 (col9), FAMILY fam3 (col4), FAMILY fam4 (col1), FAMILY fam5 (col7), FAMILY fam6 (col6), FAMILY fam7 (col12), FAMILY fam8 (col8), FAMILY fam9 (col0), FAMILY fam10 (col11), FAMILY fam11 (col3), PRIMARY KEY (col2 DESC), UNIQUE (col2 ASC, col10, col0 DESC), INDEX (col12, col2), UNIQUE (col2, col11 DESC, col9 ASC, col6, col1, col12 DESC, col10 ASC), UNIQUE (col6 ASC, col2 ASC, col8, col0, col4 DESC), UNIQUE (col10, col9 ASC, col2 DESC, col12, col7 DESC, col6 DESC), UNIQUE (col8, col7 DESC, col9 DESC, col11 ASC), UNIQUE (col4, col0 ASC, col11 DESC, col6 DESC, col2));CREATE TABLE table8 (col0 FLOAT8 NULL, col1 BYTES, col2 TIME NOT NULL, col3 NAME NULL, col4 TIMESTAMP NULL, col5 TIME NOT NULL, col6 STRING NULL, col7 BYTES, col8 UUID, col9 UUID NOT NULL, col10 TIMESTAMPTZ NOT NULL, col11 DECIMAL NULL, col12 UUID NULL, col13 TIMESTAMPTZ, col14 VARBIT NULL, FAMILY fam0 (col0), FAMILY fam1 (col7, col2), FAMILY fam2 (col9, col12), FAMILY fam3 (col6), FAMILY fam4 (col14, col5, col4, col11, col10), FAMILY fam5 (col1, col3, col8, col13), PRIMARY KEY (col0 DESC), UNIQUE (col9 ASC, col10 ASC, col7, col4 ASC, col8, col5 ASC, col3 ASC), INDEX (col14, col2 DESC, col3 DESC), UNIQUE (col4, col3 ASC, col12 DESC, col11 DESC, col7 DESC, col0 DESC, col2 DESC, col8 DESC, col13, col6 DESC, col1 DESC), INDEX (col4, col0 DESC, col11, col12 DESC, col10 ASC, col6, col1 ASC, col7, col13 DESC, col8, col3, col5 DESC));CREATE TABLE table9 (col0 STRING, col1 FLOAT8 NOT NULL, col2 TIME NOT NULL, col3 UUID NULL, col4 UUID NULL, col5 VARBIT NOT NULL, col6 VARBIT NOT NULL, col7 UUID NULL, col8 BYTES, col9 NAME NOT NULL, col10 BOOL NOT NULL, col11 UUID NOT NULL, col12 FLOAT8 NULL, col13 INTERVAL, col14 INET NOT NULL, col15 INT8[] NOT NULL, col16 DATE NOT NULL, col17 STRING NOT NULL, col18 FLOAT8 NULL, FAMILY fam0 (col13, col8, col7, col4, col15, col3, col17, col11, col16, col6, col12, col1, col2, col18, col9, col0, col10, col5, col14));
CREATE TABLE

Time: 37.684ms

root@127.0.0.1:52060/defaultdb> values (NULL, (case when true then NULL else case when false then ('\xe092542451ffb5d9' || NULL) else '\x9c82' end end || case when true then NULL else '\xa2581f03e6b23fd909' end), -0.34324108760767225, case when ('0.1940425733983187' != NULL) then NULL else '34 years 4 mons 105 days 17:24:35.110841' end, 46027248.71758977345, case when NULL then B'0111000011' else NULL end, (-5523222936787585506 + NULL), '4f65:5865:9f82:96b9:7e48:cb62:6a0e:8731/21'), ('1985-10-26', case when false then '\xc30428f888c3' else cast(coalesce(NULL, NULL) as bytea) end, -1.2105329790308406, '5 years 8 mons 348 days 00:31:26.205591', -8.554602131625729806E+27, (select B'111111010111110100011000111111000100000' from table5 as tab_3 where exists(select -0.9343659101040931, -3.787756988343020055E+26, 'eb62a51b-091d-4fc3-a220-98e494fb63f1', '16:36:04.643107', '13:34:55.201289' from table9 as tab_4 where (e'1Zj\x1e' != e'\x10\x02)~@55')) limit 1), (select '1991-03-17' from table6 as tab_5 limit 1), case when case when true then has_column_privilege(824354300, e'#C\x19\x0e\x17', (select 'Fw' from [insert into table2 (col5, col2, col3, col4, col0, col6, col7, col8, col10, col11, col14, col16, col17) select '\xd4894c7f5f45', false, B'0110100010001010100011011100000000110000101001010101000001101100101011100110110001101111001111', '54 years 11 mons 828 days 04:22:40.272428', ARRAY[1520756846664724493,-3049476035143093427,9220442524481133127], '220.247.156.15/31', '', '1970-01-03 04:58:14.000079+00:00', NULL, NULL, NULL, NULL, '90c4b1be-81d7-4b20-ac76-778dafe50577' from table2 as tab_9 limit 26 returning table2.col18, -6.269280146839308231E+32] where exists(select '5 years 11 mons 740 days 11:56:15.975738', '\x101bf8cb95bdcb63ae', NULL from table2 as tab_10 limit 60) limit 1)) else false end then '185.182.168.94/29' else NULL end), ('1980-09-03', '\x8d1a54c2fcbf5ff41d', 0.05356800740165757, case when true then '37 years 950 days 13:15:39.515493' else NULL end, (-5.52193763497928331E+35 + NULL), B'000100101010', '1981-04-01', '8c4c:e377:4ce1:39b2:1a14:6300:96db:afbe/118'), (current_date(), '\x45', -1.7234605761487996, NULL, ceiling(7.349484961592780804), B'01001011001010010000111010001110000011', current_date(), '40.173.15.218/13'), ('1978-02-25', '\x', 0.6770207242355322, (select '55 years 5 mons 870 days 17:28:42.359154' from table8 as tab_11 limit 1), NULL, case when false then NULL else NULL end, '1981-11-04', (NULL + 826652830685894045)), (current_date(), '\xf0be', -1.2008310189803173, '-22 years -3 mons -672 days -05:12:07.384984', NULL, (select B'00011010000' from table5 as tab_17 limit 1), '1993-03-23', NULL);
E190227 20:31:17.035538 411 sql/conn_executor.go:708  [n1,client=127.0.0.1:52062,user=root] a SQL panic has occurred while executing "VALUES (NULL, (CASE WHEN true THEN NULL ELSE CASE WHEN false THEN (e'\\\\xe092542451ffb5d9' || NULL) ELSE e'\\\\x9c82' END END || CASE WHEN true THEN NULL ELSE e'\\\\xa2581f03e6b23fd909' END), -0.34324108760767225, CASE WHEN ('0.1940425733983187' != NULL) THEN NULL ELSE '34 years 4 mons 105 days 17:24:35.110841' END, 46027248.71758977345, CASE WHEN NULL THEN B'0111000011' ELSE NULL END, (-5523222936787585506 + NULL), '4f65:5865:9f82:96b9:7e48:cb62:6a0e:8731/21'), ('1985-10-26', CASE WHEN false THEN e'\\\\xc30428f888c3' ELSE CAST(COALESCE(NULL, NULL) AS BYTES) END, -1.2105329790308406, '5 years 8 mons 348 days 00:31:26.205591', -8.554602131625729806E+27, (SELECT B'111111010111110100011000111111000100000' FROM table5 AS tab_3 WHERE EXISTS (SELECT -0.9343659101040931, -3.787756988343020055E+26, 'eb62a51b-091d-4fc3-a220-98e494fb63f1', '16:36:04.643107', '13:34:55.201289' FROM table9 AS tab_4 WHERE (e'1Zj\\x1e' != e'\\x10\\x02)~@55')) LIMIT 1), (SELECT '1991-03-17' FROM table6 AS tab_5 LIMIT 1), CASE WHEN CASE WHEN true THEN has_column_privilege(824354300, e'#C\\x19\\x0e\\x17', (SELECT 'Fw' FROM [INSERT INTO table2(col5, col2, col3, col4, col0, col6, col7, col8, col10, col11, col14, col16, col17) SELECT e'\\\\xd4894c7f5f45', false, B'0110100010001010100011011100000000110000101001010101000001101100101011100110110001101111001111', '54 years 11 mons 828 days 04:22:40.272428', ARRAY[1520756846664724493, -3049476035143093427, 9220442524481133127], '220.247.156.15/31', '', '1970-01-03 04:58:14.000079+00:00', NULL, NULL, NULL, NULL, '90c4b1be-81d7-4b20-ac76-778dafe50577' FROM table2 AS tab_9 LIMIT 26 RETURNING table2.col18, -6.269280146839308231E+32] WHERE EXISTS (SELECT '5 years 11 mons 740 days 11:56:15.975738', e'\\\\x101bf8cb95bdcb63ae', NULL FROM table2 AS tab_10 LIMIT 60) LIMIT 1)) ELSE false END THEN '185.182.168.94/29' ELSE NULL END), ('1980-09-03', e'\\\\x8d1a54c2fcbf5ff41d', 0.05356800740165757, CASE WHEN true THEN '37 years 950 days 13:15:39.515493' ELSE NULL END, (-5.52193763497928331E+35 + NULL), B'000100101010', '1981-04-01', '8c4c:e377:4ce1:39b2:1a14:6300:96db:afbe/118'), (current_date(), e'\\\\x45', -1.7234605761487996, NULL, ceiling(7.349484961592780804), B'01001011001010010000111010001110000011', current_date(), '40.173.15.218/13'), ('1978-02-25', e'\\\\x', 0.6770207242355322, (SELECT '55 years 5 mons 870 days 17:28:42.359154' FROM table8 AS tab_11 LIMIT 1), NULL, CASE WHEN false THEN NULL ELSE NULL END, '1981-11-04', (NULL + 826652830685894045)), (current_date(), e'\\\\xf0be', -1.2008310189803173, '-22 years -3 mons -672 days -05:12:07.384984', NULL, (SELECT B'00011010000' FROM table5 AS tab_17 LIMIT 1), '1993-03-23', NULL)": could not find overload for binary expression concat
E190227 20:31:17.038016 411 util/log/crash_reporting.go:202  [n1,client=127.0.0.1:52062,user=root] a panic has occurred!
E190227 20:31:17.046895 411 util/log/crash_reporting.go:476  [n1,client=127.0.0.1:52062,user=root] Reported as error 4d89504771d34c368d78b28d5e604d3d
panic: could not find overload for binary expression concat [recovered]
        panic: could not find overload for binary expression concat [recovered]
        panic: panic while executing 1 statements: VALUES (_, (CASE WHEN _ THEN _ ELSE CASE WHEN _ THEN (_ || _) ELSE _ END END || CASE WHEN _ THEN _ ELSE _ END), _, CASE WHEN (_ != _) THEN _ ELSE _ END, _, CASE WHEN _ THEN _ ELSE _ END, (_ + _), _), (__more5__); caused by could not find overload for binary expression concat

goroutine 411 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc0016f0000, 0x7461600, 0xc0015e49c0, 0x67b14c0, 0xc0004bcb60)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:722 +0x36d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc0016f0000, 0x7461600, 0xc0015e49c0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:427 +0x61
panic(0x67b14c0, 0xc0004bcb60)
        /usr/local/Cellar/go/1.11/libexec/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build.func1(0xc000c2c8b0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:142 +0x9a
panic(0x67b14c0, 0xc0004bcb60)
        /usr/local/Cellar/go/1.11/libexec/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.BinaryAllowsNullArgs(0xc0003c001a, 0x747ee20, 0x92cf140, 0x747ee20, 0x92cf140, 0xc00006d900)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/typing.go:80 +0x132
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).AllowNullArgs(0xc0016f0ba0, 0x92c001a, 0x7483440, 0xc0003c6940, 0x7483440, 0xc0003c6940, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/custom_funcs.go:1090 +0x93
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructConcat(0xc0016f0b90, 0x7483440, 0xc0003c6940, 0x7483440, 0xc0003c6940, 0xc000c2c001, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:12475 +0x4bd
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).constructBinary(0xc00110c000, 0xa, 0x7483440, 0xc0003c6940, 0x7483440, 0xc0003c6940, 0x747ec40, 0x92cf140, 0xc0003c6940, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:641 +0x4d8
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildScalar(0xc00110c000, 0x747be20, 0xc001841280, 0xc00153e000, 0x0, 0x0, 0x0, 0x7483440, 0xc0003c6940)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:195 +0xc7e
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildValuesClause(0xc00110c000, 0xc0003bd180, 0x0, 0x0, 0x0, 0xc00153e000, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/values.go:68 +0x361
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildSelect(0xc00110c000, 0xc0018411c0, 0x0, 0x0, 0x0, 0xc00153e000, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:587 +0x6cc
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildStmt(0xc00110c000, 0x74671c0, 0xc0018411c0, 0xc00153e000, 0x4)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:211 +0x10e
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build(0xc00110c000, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:156 +0x111
github.com/cockroachdb/cockroach/pkg/sql.(*optPlanningCtx).buildExecMemo(0xc001b2e620, 0x74616c0, 0xc001759ce0, 0xc0018411c0, 0x74616c0, 0xc001759ce0, 0x4060d00)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:399 +0x188
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc0016f0460, 0x74616c0, 0xc001759ce0, 0x6ccc29b, 0x19, 0xc4a7ec97b081f, 0x215dfe0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:147 +0xc8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).makeExecPlan(0xc0016f0000, 0x74616c0, 0xc001759ce0, 0xc0016f0460, 0x6, 0x400ecdf)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:939 +0x11b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0016f0000, 0x74616c0, 0xc001759ce0, 0xc0016f0460, 0xb9c0128, 0xc000117ad0, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:834 +0x148
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0016f0000, 0x74616c0, 0xc001759ce0, 0x74671c0, 0xc0018411c0, 0xc001361000, 0xa45, 0x0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:447 +0xd8c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0016f0000, 0x74616c0, 0xc001759ce0, 0x74671c0, 0xc0018411c0, 0xc001361000, 0xa45, 0x0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:103 +0x610
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0016f0000, 0x7461600, 0xc0015e49c0, 0xc000125e38, 0x5400, 0x15000, 0xc000125ed0, 0xc000366b90, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1183 +0x21e5
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc0006f1040, 0x7461600, 0xc0015e49c0, 0xc0016f0000, 0x5400, 0x15000, 0xc000125ed0, 0xc000366b90, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:429 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc0006f1040, 0xc000c2e000, 0xc00038cb74, 0xc000366b90, 0x7461600, 0xc0015e49c0, 0xc0016f0000, 0x5400, 0x15000, 0xc000125ed0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:331 +0xcf
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:314 +0x1040
```

cc @justinj

--- a/pkg/sql/opt/memo/typing.go
+++ b/pkg/sql/opt/memo/typing.go
@@ -65,6 +65,25 @@ func InferBinaryType(op opt.Operator, leftType, rightType types.T) types.T {
 	return o.ReturnType
 }
 
+// InferWhensType returns the type of a CASE expression, which is
+// of the form:
+//   CASE [ <cond> ]
+//       WHEN <condval1> THEN <expr1>
+//     [ WHEN <condval2> THEN <expr2> ] ...
+//     [ ELSE <expr> ]
+//   END
+// The type is equal to the type of the WHEN <condval> THEN <expr> clauses, or
+// the type of the ELSE <expr> value if all the previous types are unknown.
+func InferWhensType(whens ScalarListExpr, orElse opt.ScalarExpr) types.T {
+	for _, when := range whens {
+		childType := when.DataType()
+		if childType != types.Unknown {
+			return childType
+		}
+	}
+	return orElse.DataType()
+}
+
 // BinaryOverloadExists returns true if the given binary operator exists with the
 // given arguments.
 func BinaryOverloadExists(op opt.Operator, leftType, rightType types.T) bool {
@@ -280,13 +299,7 @@ func typeCoalesce(e opt.ScalarExpr) types.T {
 // the type of the ELSE <expr> value if all the previous types are unknown.
 func typeCase(e opt.ScalarExpr) types.T {
 	caseExpr := e.(*CaseExpr)
-	for _, when := range caseExpr.Whens {
-		childType := when.DataType()
-		if childType != types.Unknown {
-			return childType
-		}
-	}
-	return caseExpr.OrElse.DataType()
+	return InferWhensType(caseExpr.Whens, caseExpr.OrElse)
 }
 
 // typeWhen returns the type of a WHEN <condval> THEN <expr> clause inside a

--- a/pkg/sql/opt/norm/custom_funcs.go
+++ b/pkg/sql/opt/norm/custom_funcs.go
@@ -1147,28 +1147,15 @@ func (c *CustomFuncs) IsConstValueEqual(const1, const2 opt.ScalarExpr) bool {
 	}
 }
 
-// ensureTyped makes sure that any NULL passing through gets tagged with an
-// appropriate type.
-func (c *CustomFuncs) ensureTyped(d opt.ScalarExpr, typ types.T) opt.ScalarExpr {
-	if d.DataType() == types.Unknown {
-		return c.f.ConstructNull(typ)
-	}
-	return d
-}
-
 // SimplifyWhens removes known unreachable WHEN cases and constructs a new CASE
 // statement. Any known true condition is converted to the ELSE. If only the
 // ELSE remains, its expression is returned. condition must be a ConstValue.
 func (c *CustomFuncs) SimplifyWhens(
 	condition opt.ScalarExpr, whens memo.ScalarListExpr, orElse opt.ScalarExpr,
 ) opt.ScalarExpr {
-	var typ types.T
 	newWhens := make(memo.ScalarListExpr, 0, len(whens))
 	for _, item := range whens {
 		when := item.(*memo.WhenExpr)
-		if typ == nil || typ == types.Unknown {
-			typ = when.Typ
-		}
 		if opt.IsConstValueOp(when.Condition) {
 			if !c.IsConstValueEqual(condition, when.Condition) {
 				// Ignore known unmatching conditions.
@@ -1178,7 +1165,7 @@ func (c *CustomFuncs) SimplifyWhens(
 			// If this is true, we won't ever match anything else, so convert this to
 			// the ELSE (or just return it if there are no earlier items).
 			if len(newWhens) == 0 {
-				return when.Value
+				return c.ensureTyped(when.Value, memo.InferWhensType(whens, orElse))
 			}
 			return c.f.ConstructCase(condition, newWhens, when.Value)
 		}
@@ -1193,10 +1180,19 @@ func (c *CustomFuncs) SimplifyWhens(
 		// a type we observed earlier.
 		// typ will never be nil here because the definition of
 		// SimplifyCaseWhenConstValue ensures that whens is nonempty.
-		return c.ensureTyped(orElse, typ)
+		return c.ensureTyped(orElse, memo.InferWhensType(whens, orElse))
 	}
 
 	return c.f.ConstructCase(condition, newWhens, orElse)
+}
+
+// ensureTyped makes sure that any NULL passing through gets tagged with an
+// appropriate type.
+func (c *CustomFuncs) ensureTyped(d opt.ScalarExpr, typ types.T) opt.ScalarExpr {
+	if d.DataType() == types.Unknown {
+		return c.f.ConstructNull(typ)
+	}
+	return d
 }
 
 // OpsAreSame returns true if the two operators are the same.

--- a/pkg/sql/opt/norm/rules/scalar.opt
+++ b/pkg/sql/opt/norm/rules/scalar.opt
@@ -312,8 +312,8 @@
     $cmp
 )
 
-# FoldCollate converts a Collate expr over an uncollated string into a collated string
-# constant.
+# FoldCollate converts a Collate expr over an uncollated string into a collated
+# string constant.
 [FoldCollate, Normalize]
 (Collate
     $input:(Const)

--- a/pkg/sql/opt/norm/testdata/rules/scalar
+++ b/pkg/sql/opt/norm/testdata/rules/scalar
@@ -618,7 +618,7 @@ values
  └── (NULL,) [type=tuple{string}]
 
 # Regression test for #34930.
-norm
+norm expect=SimplifyCaseWhenConstValue
 SELECT
     CASE
     WHEN true THEN NULL
@@ -632,6 +632,19 @@ values
  ├── key: ()
  ├── fd: ()-->(1)
  └── (NULL,) [type=tuple{decimal}]
+
+# Regression test for #35246.
+norm expect=SimplifyCaseWhenConstValue
+SELECT
+    CASE WHEN true THEN NULL ELSE 'foo' END ||
+    CASE WHEN true THEN NULL ELSE 'bar' END
+----
+values
+ ├── columns: "?column?":1(string)
+ ├── cardinality: [1 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1)
+ └── (NULL,) [type=tuple{string}]
 
 # Verify that a true condition does not remove non-constant expressions
 # proceeding it.

